### PR TITLE
Fix unsafe yield from rewrite

### LINKED ISSUE #216: 'yield from' rewrite unsafe when looping variable is referenced later
Consider the example code:

```py
def foo(it):
    item = None
    for item in it:
        yield item
    print("Last item", item)
```

With `--py3-plus`, this is transformed to:

```py
def foo(it):
    item = None
    yield from it
    print("Last item", item)
```

But these are not equivalent. The line `print("Last item", item)` now always prints `None` rather than the last item of the passed iterable. 

This was discovered in more-itertools:

https://github.com/erikrose/more-itertools/blob/8d860c355a1dd900e3a9c5fe1303ddf5a778452f/more_itertools/more.py#L1206-L1223

### PR DESCRIPTION
**Pull request by atugushev**
Fixes #216.

Prevents rewrite `for-loop yield` if a loop variable referenced after the loop.

Also, it fixes along the way the case with yield outside the function. Now `pyupgrade` doesn't rewrite it:
```bash
# before
$ echo "for x in y: yield x" | pyupgrade - --py3-plus
yield from y

# after
$ echo "for x in y: yield x" | pyupgrade - --py3-plus
for x in y: yield x
```
<sup>I don't know is it intentional behavior or not. Please, inform if it is.</sup>

Additionally, I've added a test for `lambda yield` (weird but syntactically possible), since there is `Union[ast.FunctionDef, ast.Lambda]`.

### CODE REVIEW

**Comment by atugushev** [comment]
Aw, sorry. I've forgotten to check the coverage.

File: pyupgrade.py [hunk 1]
```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
         super(FindPy3Plus, self).generic_visit(node)
 
 
+def reversed_walk(node):  # type: (ast.AST) -> Generator[ast.AST, None, None]
+    todo = [node]
+    while todo:
```
Comment (atugushev) [inline pyupgrade.py]: @asottile would it be okay to add `pragma: no branch` (for it's `while True`) here and further walk-branches? Or it should be refactored?

File: pyupgrade.py [hunk 2]
```diff
@@ -1307,19 +1308,13 @@ def _visit_func(self, node):
...
         self._visit_func(node)
-        self._in_async_def = orig
+        self._current_sync_def = orig
 
     visit_FunctionDef = visit_Lambda = _visit_sync_func
 
-    def visit_AsyncFunctionDef(self, node):  # pragma: no cover (py35+)
-        # type: (AsyncFunctionDef) -> None
-        self._in_async_def, orig = True, self._in_async_def
-        self._visit_func(node)
```
Comment (asottile) [inline pyupgrade.py]: this introduces a bug, can you write a test for old `super` inside an `async def`?

File: pyupgrade.py (hunk 1 above)
Comment (asottile) [reply pyupgrade.py]: you should be able to hit the coverage here -- the missing branch indicates this either never terminates or never loops more than once

File: pyupgrade.py [hunk 3]
```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
...
+    while todo:
+        node = todo.pop()
+        todo.extend(ast.iter_child_nodes(node))
+        yield node
+
+
+def targets_referenced_after_node(root_node, node, targets):
+    # type: (ast.AST, ast.AST, Union[ast.expr, ast.Tuple]) -> bool
+    """
+    Search inside the given root node whether the targets have been referenced
```
Comment (asottile) [inline pyupgrade.py]: docstrings should not start with whitespace

File: pyupgrade.py [hunk 4]
```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
...
 
+def reversed_walk(node):  # type: (ast.AST) -> Generator[ast.AST, None, None]
+    todo = [node]
+    while todo:
+        node = todo.pop()
+        todo.extend(ast.iter_child_nodes(node))
+        yield node
+
+
+def targets_referenced_after_node(root_node, node, targets):
```
Comment (asottile) [inline pyupgrade.py]: let's underscore these functions, pyupgrade has no programmatic api

File: pyupgrade.py [hunk 5]
```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
...
+    # type: (ast.AST, ast.AST, Union[ast.expr, ast.Tuple]) -> bool
+    """
+    Search inside the given root node whether the targets have been referenced
+    after the given node. Walks from behind until the node have been reached.
+    """
+    for x in reversed_walk(root_node):
```
Comment (asottile) [inline pyupgrade.py]: alternatively you forward walk (don't need the other function) and then you start looking when the node is encountered

File: pyupgrade.py [hunk 6]
```diff
@@ -1520,12 +1515,15 @@ def visit_If(self, node):  # type: (ast.If) -> None
...
-            not self._in_async_def and
+            self._current_sync_def is not None and
             len(node.body) == 1 and
             isinstance(node.body[0], ast.Expr) and
             isinstance(node.body[0].value, ast.Yield) and
             node.body[0].value.value is not None and
             targets_same(node.target, node.body[0].value.value) and
+            not targets_referenced_after_node(
+                self._current_sync_def, node, node.target,
+            ) and
```
Comment (asottile) [inline pyupgrade.py]: not a big deal but if you could shorten this to fit on one line that'd be great

File: tests/yield_from_test.py [hunk 7]
```diff
@@ -184,6 +184,48 @@ def test_fix_async_yield_from(s, expected):
         '        yield x\n'
         '    else:\n'
         '        print("boom!")\n',
+        pytest.param(
+            'for x in z:\n'
+            '    yield x\n',
+            id='yield outside function',
+        ),
+        pytest.param(
+            'lambda: [(yield x) for x in z]',
```
Comment (asottile) [inline tests/yield_from_test.py]: I don't think this test is relevant, there's no `for` loop here

File: pyupgrade.py [hunk 8]
```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
...
+            break
+
+        # Filter for name expressions
+        if not isinstance(x, ast.Name):
+            continue
+
+        # Match one of the targets
+        for t in ast.walk(targets):
+            if fields_same(t, x):
+                return True
```
Comment (asottile) [inline pyupgrade.py]: It would probably be better to refactor to this (above the loop)

```python
names = {node.id for node in ast.walk(targets)}

...

if isinstance(x, ast.Name) and isinstance(x.ctx, ast.Load) and x.id in names:
    ...
```

Comment (atugushev) [reply pyupgrade.py]:
```python
names = {node.id for node in ast.walk(targets)}
```
That's neat, but it seems it doesn't handle targets as `ast.Tuple` (`for x, y in z: yield x, y` case). That's why there's `fields_same`.

Comment (atugushev) [reply pyupgrade.py]: Maybe, it's better to walk through the targets and compose the set with node.id.

Comment (asottile) [reply pyupgrade.py]: it's missing an `if isinstance(node, ast.Name)` -- that's what I get for phone

Comment (atugushev) [reply pyupgrade.py]: Yay. that's working!

File: pyupgrade.py (hunk 6 above)
Comment (atugushev) [reply pyupgrade.py]: The naming and signature are okay?

Comment (asottile) [reply pyupgrade.py]: yeah I meant if you can come up with more succinct names such that it fits, not a big deal

File: pyupgrade.py (hunk 5 above)
Comment (atugushev) [reply pyupgrade.py]: I came up with:
```python
todo = collections.deqeu(func_node.body)
while todo:
    node = todo.popleft()
    ....
```
What do you think?

Comment (asottile) [reply pyupgrade.py]: I'd prefer trying to make it work with `ast.walk` if possible -- failing that whatever works

Comment (atugushev) [reply pyupgrade.py]: Can't figure out how to achieve the same with `ast.walk`

```python
    visited = False
    todo = collections.deque(func_node.body)

    while todo:
        node = todo.popleft()

        if node == forloop_node:
            visited = True
            continue

        if not visited:
            continue

        if (
            isinstance(node, ast.Name) and
            isinstance(node.ctx, ast.Load) and
            node.id in names
        ):
            return True

        todo.extend(ast.iter_child_nodes(node))

    return False
```

How to prevent walking inside the for loop? Have any ideas?

Comment (atugushev) [reply pyupgrade.py]: Dammit, I'm too bad at walking trees 😒

Comment (asottile) [reply pyupgrade.py]: adjust the `visited` assignment until the last element of the `yield` expression maybe

Comment (atugushev) [reply pyupgrade.py]: Too complicated, maybe node visitor would help? I'll try that.

Comment (asottile) [reply pyupgrade.py]: if you're using a node visitor you could reuse the existing walk and remove the loops that are referenced after -- this would also keep it 1-pass :) (you might need to store some additional state, but you could use `visit_Name` directly)

Comment (atugushev) [reply pyupgrade.py]: Strange just happened. I wrote the visitor and tests passed :D And it's less code.

Comment (atugushev) [reply pyupgrade.py]: Just a thought, can we skip the nodes till for loop depend on offset information?

Comment (asottile) [reply pyupgrade.py]: not always, for instance

```python
f'''
{foo}
'''
```

reports an offset of -1 on python<3.8

File: pyupgrade.py (hunk 1 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.

File: pyupgrade.py (hunk 3 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.

File: pyupgrade.py (hunk 4 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.

File: pyupgrade.py (hunk 5 above)
Comment (atugushev) [reply pyupgrade.py]: Please review in 81dc036f6a52e32a3dd1e992334c24fc09cea12b  how it's done via visitor and stack depth level checking.

File: pyupgrade.py (hunk 6 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.

File: pyupgrade.py (hunk 8 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 81dc036f6a52e32a3dd1e992334c24fc09cea12b.

File: tests/yield_from_test.py (hunk 7 above)
Comment (atugushev) [reply tests/yield_from_test.py]: Addressed in 9572c5a7dadf6a4f8c5877462aa23d05be9168fc.

File: pyupgrade.py (hunk 2 above)
Comment (atugushev) [reply pyupgrade.py]: Sorry, I'm afraid I'm not following. Could you elaborate on this?

Comment (atugushev) [reply pyupgrade.py]: Should I revert back this block and restore `if not self._in_async_def` in `visit_For`?

Comment (asottile) [reply pyupgrade.py]: this fails rewrite now:

```python
class C:
    async def foo(self):
        super(C, self).foo()
```

just needs a quick test for that and to make sure this still calls `_visit_func` essentially

Comment (atugushev) [reply pyupgrade.py]: I see. Thanks for noticing!

Comment (atugushev) [reply pyupgrade.py]: Addressed in 302ab16283facc67a6cc5ab189af10ba915ae05b.

File: pyupgrade.py (hunk 5 above)
Comment (asottile) [reply pyupgrade.py]: cool I'll poke at this, might be another way to approach it I'll see what I can do 👍

Comment (atugushev) [reply pyupgrade.py]: Thank you so much!

Comment (atugushev) [reply pyupgrade.py]: > adjust the `visited` assignment until the last element of the `yield` expression maybe

FTR, looks like it's impossible since `ast.walk` does undefined-order traversal.

**Comment by asottile** [comment]
can you add this testcase (should be a noop):

```python
def f():
    async def g():
        for y in z:
            yield y
    return g
```

**Comment by asottile** [comment]
ok had another chance to look at this:

```diff
diff --git a/pyupgrade.py b/pyupgrade.py
index 8cd3a13..2f4ec67 100644
--- a/pyupgrade.py
+++ b/pyupgrade.py
@@ -1197,7 +1197,8 @@ class FindPy3Plus(ast.NodeVisitor):
         self._class_info_stack = []  # type: List[FindPy3Plus.ClassInfo]
         self._in_comp = 0
         self.super_calls = {}  # type: Dict[Offset, ast.Call]
-        self._current_func_node = None  # type: Optional[SyncFunctionDef]
+        self._current_func = None  # type: Optional[SyncFunctionDef]
+        self._for_targets = {}  # type: Dict[Tuple[SyncFunctionDef, str], Offset]  # noqa: E501
         self.yield_from_fors = set()  # type: Set[Offset]
 
     def _is_six(self, node, names):
@@ -1309,9 +1310,9 @@ class FindPy3Plus(ast.NodeVisitor):
 
     def _visit_sync_func(self, node):
         # type: (SyncFunctionDef) -> None
-        self._current_func_node, orig = node, self._current_func_node
+        self._current_func, orig = node, self._current_func
         self._visit_func(node)
-        self._current_func_node = orig
+        self._current_func = orig
 
     visit_FunctionDef = visit_Lambda = _visit_sync_func
 
@@ -1518,17 +1519,28 @@ class FindPy3Plus(ast.NodeVisitor):
         self.generic_visit(node)
 
     def visit_For(self, node):  # type: (ast.For) -> None
+        self.generic_visit(node)
+        # done *after* `generic_visit` so we know all things after this are
+        # after this `for` loop
         if (
-            self._current_func_node is not None and
+            self._current_func is not None and
             len(node.body) == 1 and
             isinstance(node.body[0], ast.Expr) and
             isinstance(node.body[0].value, ast.Yield) and
             node.body[0].value.value is not None and
             targets_same(node.target, node.body[0].value.value) and
-            not _targets_referenced(self._current_func_node, node) and
             not node.orelse
         ):
-            self.yield_from_fors.add(_ast_to_offset(node))
+            offset = _ast_to_offset(node)
+            self.yield_from_fors.add(offset)
+            for node in ast.walk(node.target):
+                if isinstance(node, ast.Name):
+                    self._for_targets[(self._current_func, node.id)] = offset
+
+    def visit_Name(self, node):  # type: (ast.Name) -> None
+        offset = self._for_targets.get((self._current_func, node.id))
+        if offset is not None:
+            self.yield_from_fors.discard(offset)
 
         self.generic_visit(node)
 
@@ -1537,48 +1549,6 @@ class FindPy3Plus(ast.NodeVisitor):
         super(FindPy3Plus, self).generic_visit(node)
 
 
-class FindNamesRefsAfterForLoopVisitor(ast.NodeVisitor):
-    def __init__(self, forloop):  # type: (ast.For) -> None
-        self.forloop = forloop
-
-        self._stack_depth = 0
-        self._forloop_depth = None  # type: Optional[int]
-        self.found = set()  # type: Set[str]
-
-    def visit_For(self, node):  # type: (ast.For) -> None
-        if node == self.forloop:
-            self._forloop_depth = self._stack_depth
-        else:
-            self.generic_visit(node)
-
-    def visit_Name(self, node):  # type: (ast.Name) -> None
-        if (
-            isinstance(node.ctx, ast.Load) and
-            self._forloop_depth is not None and
-            self._stack_depth > self._forloop_depth
-        ):
-            self.found.add(node.id)
-        self.generic_visit(node)
-
-    def visit(self, node):  # type: (ast.AST) -> None
-        super(FindNamesRefsAfterForLoopVisitor, self).visit(node)
-        self._stack_depth += 1
-
-
-def _targets_referenced(func, forloop):
-    # type: (SyncFunctionDef, ast.For) -> bool
-    """Search inside the given function node whether the loop variables
-    have been referenced after the loop.
-    """
-    visitor = FindNamesRefsAfterForLoopVisitor(forloop)
-    visitor.visit(func)
-    names = {
-        node.id for node in ast.walk(forloop.target)
-        if isinstance(node, ast.Name)
-    }
-    return len(visitor.found & names) > 0
-
-
 def _fixup_dedent_tokens(tokens):  # type: (List[Token]) -> None
     """For whatever reason the DEDENT / UNIMPORTANT_WS tokens are misordered
 
```

let me know what you think of that

also probably needs a test for:

```python
def f():
    for x in y:
        yield x
    x += 1
```

I think this uses `ctx=Store()` despite being an augmented store :S

**Comment by atugushev** [comment]
> let me know what you think of that

That's brilliant!

**Comment by atugushev** [comment]
I've applied your patch in 424410f and added tests and a few changes to make mypy happy in a083dcf.

File: pyupgrade.py [hunk 9]
```diff
@@ -1295,7 +1301,8 @@ def visit_ClassDef(self, node):  # type: (ast.ClassDef) -> None
         self._class_info_stack.pop()
 
     def _visit_func(self, node):
-        # type: (Union[ast.FunctionDef, ast.Lambda, AsyncFunctionDef]) -> None
+        # type: (SyncOrAsyncFunctionDef) -> None
+        self._current_func, orig = node, self._current_func
```
Comment (asottile) [inline pyupgrade.py]: instead you can probably set `_current_func` to `None` inside `visit_AsyncFunctionDef`

File: tests/super_test.py [hunk 10]
```diff
@@ -110,6 +112,19 @@ def test_fix_super_noop(s):
...
+            '        super(C, self).foo()\n',
+            'class C:\n'
+            '    async def foo(self):\n'
+            '        super().foo()\n',
+            id='super inside async func def',
+            marks=pytest.mark.xfail(
+                sys.version_info < (3, 5),
+                reason='async introduced in python 3.5',
+            ),
+        ),
```
Comment (asottile) [inline tests/super_test.py]: let's split this into a separate function

File: pyupgrade.py [hunk 11]
```diff
@@ -58,6 +58,11 @@
         AsyncFunctionDef = ast.AsyncFunctionDef
     else:
         AsyncFunctionDef = ast.stmt
+    SyncOrAsyncFunctionDef = Union[
+        ast.AsyncFunctionDef,
+        ast.FunctionDef,
+        ast.Lambda
```
Comment (asottile) [inline pyupgrade.py]: if you upgrade `add-trailing-comma` to the latest version it will fix this

File: pyupgrade.py (hunk 9 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in a070506d51ca8e715a83912e6adaa8f6fbcc0b85.

File: tests/super_test.py (hunk 10 above)
Comment (atugushev) [reply tests/super_test.py]: Addressed in 626fead0f4bba195eed39bc5ee9b1b9612db4bfc.

File: pyupgrade.py (hunk 11 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 7f7709432965ef668110d1fe96f932b9f02d3a88.

File: tests/super_test.py [hunk 12]
```diff
@@ -129,3 +129,25 @@ def test_fix_super_noop(s):
...
+            'class C:\n'
+            '    async def foo(self):\n'
+            '        super(C, self).foo()\n',
+            'class C:\n'
+            '    async def foo(self):\n'
+            '        super().foo()\n',
+            id='super inside async func def',
+        ),
+    ),
+)
```
Comment (asottile) [inline tests/super_test.py]: probably don't need parametrize if there's only one test

File: pyupgrade.py [hunk 13]
```diff
@@ -1196,7 +1201,8 @@ def __init__(self):  # type: () -> None
         self._class_info_stack = []  # type: List[FindPy3Plus.ClassInfo]
         self._in_comp = 0
         self.super_calls = {}  # type: Dict[Offset, ast.Call]
-        self._in_async_def = False
+        self._current_func = None  # type: Optional[SyncOrAsyncFunctionDef]
```
Comment (asottile) [inline pyupgrade.py]: this type is now overly broad here (I'd probably remove `SyncOrAsyncFunctionDef` as well)

Comment (atugushev) [reply pyupgrade.py]: Would you like to change it to
```python
self._current_func = None  # type: Optional[Union[ast.AsyncFunctionDef, ast.FunctionDef, ast.Lambda]]  # noqa: E501
```

instead?

File: tests/super_test.py (hunk 12 above)
Comment (atugushev) [reply tests/super_test.py]: Yeah, right. I've forgotten to remove the old test btw 🤦‍♂

File: pyupgrade.py (hunk 13 above)
Comment (asottile) [reply pyupgrade.py]: you changed it so it can no longer contain `AsyncFunctionDef`

**Comment by asottile** [comment]
hmm shoot I just thought of another case :/

```python
def f():
    def g():
        return x

    for x in y:
        yield x
    yield g()
```

this complicates things significantly I think :/

File: pyupgrade.py (hunk 13 above)
Comment (atugushev) [reply pyupgrade.py]: Got it!

**Comment by atugushev** [comment]
Ouch!

cat

**Comment by asottile** [comment]
so the adjusted logic probably needs to be something like: accessed anywhere outside of the loop body :/

probably needs a stack of nested functions and to store referenced names in them (and avoid it in the loop body)

this is much more complicated than initially thought lol

File: tests/super_test.py (hunk 12 above)
Comment (atugushev) [reply tests/super_test.py]: Addressed in 56e91882b1c626d72c97882f0352d21ae93ae0fb.

File: pyupgrade.py (hunk 13 above)
Comment (atugushev) [reply pyupgrade.py]: Addressed in 56e91882b1c626d72c97882f0352d21ae93ae0fb.

**Comment by asottile** [comment]
via #286
